README.md

```diff
@@ -1,10 +1,148 @@
+         @crossmint/ solana-auth-react-ui@beta
+```
+
+Define `signIn` and `signOut`. Here we use firebase utils and `signInWithCustomToken`.
+
+```js
+// utils/firebaseClient.ts
+export async function signIn(data: Record<string, string>) {
+  const userCredential = await signInWithCustomToken(auth, data.token);
+  return userCredential;
+}
+
+export const signOut = () => {
+  return auth.signOut();
+};
+```
+
+Wrap your app in `SolanaAuthProvider`. Make sure to pass it a domain, your auth fucntions, and your serverless endpoints.
+
+```javascript
+// _app.tsx
+import { AppProps } from "next/app";
+import { FC } from "react";
+import { SolanaAuthProvider } from "../src";
+import { signIn, signOut } from "../utils/firebaseClient";
+
+import { SolanaAuthProvider } from "@crossmint/solana-auth-react-ui";
+
+require("../styles/globals.css");
+
+const App: FC<AppProps> = ({ Component, pageProps }) => {
+  return (
+    <SolanaAuthProvider
+      domain="example.xyz"
+      requestUrl="/api/solana-auth/getauthchallenge"
+      callbackUrl="/api/solana-auth/completeauthchallenge"
+      onAuthCallback={signIn}
+      signOut={signOut}
+    >
+      <Component {...pageProps} />
+    </SolanaAuthProvider>
+  );
+};
+
+export default App;
+```
+
+Use the `SolanaFirebaseAuth` component to render the signin button in your app.
+
+```js
+import { SolanaAuthButtonFirebase } from "@crossmint/solana-auth-react-ui";
+import { auth } from "../utils/firebaseClient";
+const Index = () => {
+  return (
+    <div>
+      <head>
+        <title>Sign in with Solana</title>
+        <meta name="description" content="Sign in with Solana" />
+        <link rel="icon" href="/favicon.ico" />
+      </head>
+
+      <main>
+        <SolanaAuthButtonFirebase auth={auth} />
+      </main>
+    </div>
+  );
+};
+
+export default Index;
+```
+
+## Use the serverless functions.
+
+```js
+import { FirebaseAdapter, SolanaAuth } from "@crossmint/solana-auth-base";
+import firebaseApp from "../../../utils/firebase";
+
+export default SolanaAuth({ adapter: FirebaseAdapter(firebaseApp) });
+```
+
+This object exposes `getAuthChallenge` and `completeAuthChallenge`. These functions take a request and response like vercel or firebase functions and return `Promise<void>`.
+
+## Client Usage
+
+export default Index;
+
+````
+
+# Custom Configuration
+
+## Defining an adapter
+
+**If you are using Firebase, we have built a database adapter for you, accesible with `@crossmint/solana-auth-base`**
+
+Database adapters must fit the `Adapter` shape
+
+```js
+type SignInAttempt = {
+  nonce: string,
+  ttl: number,
+  pubkey: string,
+};
+
+interface Adapter {
+  saveSigninAttempt(attempt: SignInAttempt): Promise<void>;
+  getNonce(pubkey: string): Promise<any>;
+  generateToken(pubkey: string): Promise<string>;
+  getTLL(pubkey: string): Promise<number>;
+}
+````
+
+A custom database adapter can be passed to the `SolanaAuth` constructor like so:
+
+```js
+SolanaAuth({ adapter: CustomAdapter() });
```

Comment:
I would add an example declaration of a custom adapter. For ex:

class CustomAdapter implements Adapter {
  saveSigninAttempt(attempt: SignInAttempt): Promise<void> {
    // TODO: save attempt object into persistent storage. Use the <public key> as primary key 
    // and replace any existing entry with the same key
  }
  getNonce(pubkey: string): Promise<any> {
    // TODO: retrieve the nonce of the most recent sign-in attempt for a public key
  }
  generateToken(pubkey: string): Promise<string> {
    // TODO (Ben) explain this one
  }
  getTLL(pubkey: string): Promise<number> {
      // TODO (Ben) explain this one
  }
}